src/solvers/refinement/string_refinement.cpp

```diff
@@ -844,14 +863,17 @@ exprt string_refinementt::substitute_array_with_expr(
     const exprt &then_expr=with_expr.new_value();
     exprt else_expr=substitute_array_with_expr(with_expr.old(), index);
     const typet &type=then_expr.type();
-    assert(else_expr.type()==type);
+    CHECK_RETURN(else_expr.type()==type);
     return if_exprt(
       equal_exprt(index, with_expr.where()), then_expr, else_expr, type);
   }
   else
   {
-    // Only handle 'with' expressions on 'array_of' expressions.
-    assert(expr.id()==ID_array_of);
+    // Only handle 'with' expressions and 'array_of' expressions.
+    INVARIANT(
```

Comment:
Not necessarily I feel. The type passed in is a `exprt`, which has pretty much no data invariants. This is more of an adhoc precondition for the function about the `exprt` it gets. In my mind the data invariant is reserved for invariants about types.